x-pack/metricbeat/module/gcp/carbon/carbon.go

```diff
@@ -0,0 +1,320 @@
+func (c config) Validate() error {
+	if c.CredentialsFilePath == "" && c.CredentialsJSON == "" {
+		return errors.New("no credentials_file_path or credentials_json specified")
+	}
+
+	if c.Period.Hours() < 24 {
+		return fmt.Errorf("collection period for carbon footprint metricset %s cannot be less than 24 hours", c.Period)
+	}
+	return nil
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The gcp '%s' metricset is beta.", metricsetName)
+
+	m := &MetricSet{
+		BaseMetricSet: base,
+		logger:        logp.NewLogger(metricsetName),
+	}
+
+	if err := base.Module().UnpackConfig(&m.config); err != nil {
+		return nil, fmt.Errorf("unpack carbon footprint config failed: %w", err)
+	}
+
+	m.Logger().Debugf("metricset config: %v", m.config)
+	return m, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) (err error) {
+	// find current month
+	month := getReportMonth(time.Now())
+
+	var opt []option.ClientOption
+	if m.config.CredentialsFilePath != "" && m.config.CredentialsJSON != "" {
+		return errors.New("both credentials_file_path and credentials_json specified, you must use only one of them")
+	} else if m.config.CredentialsFilePath != "" {
+		opt = []option.ClientOption{option.WithCredentialsFile(m.config.CredentialsFilePath)}
+	} else if m.config.CredentialsJSON != "" {
+		opt = []option.ClientOption{option.WithCredentialsJSON([]byte(m.config.CredentialsJSON))}
+	} else {
+		return errors.New("no credentials_file_path or credentials_json specified")
+	}
+
+	client, err := bigquery.NewClient(ctx, m.config.ProjectID, opt...)
+	if err != nil {
+		return fmt.Errorf("gerror creating bigquery client: %w", err)
+	}
+
+	defer client.Close()
+
+	if m.config.TableName == "" {
+		m.logger.Warn("table_name is not set in config, \"carbon_footprint\" will be used by default.")
+		m.config.TableName = "carbon_footprint"
+	}
+
+	tableMeta, err := getTable(ctx, client, m.config.DatasetID, m.config.TableName)
+	if err != nil {
+		return fmt.Errorf("getTables failed: %w", err)
+	}
+
+	var events []mb.Event
+	eventsPerQuery, err := m.queryBigQuery(ctx, client, tableMeta, month)
+	if err != nil {
+		return fmt.Errorf("queryBigQuery failed: %w", err)
+	}
+
+	events = append(events, eventsPerQuery...)
+
+	m.Logger().Debugf("Total %d of events are created for carbon footprint", len(events))
+	for _, event := range events {
+		reporter.Event(event)
+	}
+	return nil
+}
+
+// getReportMonth gets the year-month of the latest expected report.
+// GCP creates new reports on the 15 of each month. So if the date is below
```

Comment:
If I correctly understood this line, you should expect 1 report per month, on the 15th. Given this, does it make sense to collect these data with a period less than 1 month?